test/hotspot/jtreg/testlibrary_tests/template_framework/examples/TestTutorial.java

```diff
@@ -0,0 +1,539 @@
+        comp.invoke("p.xyz.InnerTest4", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest5", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest6", "main", new Object[] {});
+        comp.invoke("p.xyz.InnerTest7", "main", new Object[] {});
+    }
+
+    // This example shows the use of various Tokens.
+    public static String generateWithListOfTokens() {
+        // A Template is essencially a function / lambda that produces a
+        // token body, which is a list of Tokens that are concatenated.
+        var templateClass = Template.make(() -> body(
+            // The "body" method is filled by a sequence of "Tokens".
+            // This can be Strings and multi-line Strings, but also
+            // boxed primitives.
+            """
+            package p.xyz;
+
+            public class InnerTest1 {
+                public static void main() {
+                    System.out.println("Hello World!");
+            """,
+            "int a = ", Integer.valueOf(1), ";\n",
+            "float b = ", Float.valueOf(1.5f), ";\n",
+            // Special Float values are "smartly" formatted!
+            "float nan = ", Float.valueOf(Float.POSITIVE_INFINITY), ";\n",
+            "boolean c = ", Boolean.valueOf(true), ";\n",
+            // Lists of Tokens are also allowed:
+            List.of("int ", "d = 5", ";\n"),
+            // That can be great for streaming / mapping over an existing list:
+            List.of(3, 5, 7, 11).stream().map(i -> "System.out.println(" + i + ");\n").toList(),
+            """
+                    System.out.println(a + " " + b + " " + nan + " " + c + " " + d);
+                }
+            }
+            """
+        ));
+
+        // Render templateClass to String.
+        return templateClass.withArgs().render();
+    }
+
+    // This example shows the use of Templates, with and without arguments.
+    public static String generateWithTemplateArguments() {
+        // A Template with no arguments.
+        var templateHello = Template.make(() -> body(
+            """
+            System.out.println("Hello");
+            """
+        ));
+
+        // A Template with a single Integer argument.
+        var templateCompare = Template.make("arg", (Integer arg) -> body(
+            "System.out.println(", arg, ");\n",  // capture arg via lambda argument
+            "System.out.println(#arg);\n",       // capture arg via hashtag replacement
+            "if (#arg != ", arg, ") { throw new RuntimeException(\"mismatch\"); }\n"
+        ));
+
+        // A Template that creates the body of the Class and main method, and then
+        // uses the two Templates above inside it.
+        var templateClass = Template.make(() -> body(
+            """
+            package p.xyz;
+
+            public class InnerTest2 {
+                public static void main() {
+            """,
+                    templateHello.withArgs(),
```

Comment:
`withArgs()` looks strange when there are no args. Could we find a better name for it? But maybe I'm missing a pattern here.

Reply:
Hmm, yeah, that is a slight concern. But it does return a `TemplateWithArgs`, which means a template that knows all the arguments already. This one happens to be a zero-arg version.

I suppose I could rename it to `withArgsNone()` or `withZeroArgs` or `withNoArgs` for the zero-args version? Would that be an improvement?

Reply:
I will do the refactoring we discussed offline, with `UnfilledTemplate` and `FilledTemplate`. And `Template.make` with zero args directly generates a FilledTemplate.